radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/substatehash/BerkeleySubstateAccumulatorHashStore.java

```diff
@@ -95,66 +97,101 @@ public class BerkeleySubstateAccumulatorHashStore implements BerkeleyAdditionalS
 
   private static final Logger logger = LogManager.getLogger();
 
-  private static final String EPOCH_HASH_FILE_PATH =
-      "radixdlt-core/radixdlt/src/main/resources/epoch-hash";
-  public static final String EPOCH_HASH_FILE_SEPARATOR = "=";
+  public static final String UPDATE_EPOCH_HASH_FILE_ENABLE_PROPERTY_NAME =
+      "update_epoch_hash_file.enable";
+  public static final String VERIFY_EPOCH_HASH_ENABLE_PROPERTY_NAME = "verify_epoch_hash.enable";
+
+  private static final String EPOCH_HASH_FILE_WRITE_PATH =
+      String.join(
+          File.separator,
+          System.getProperty("user.dir"),
+          "radixdlt-core",
+          "radixdlt",
+          "src",
+          "main",
+          "resources",
+          "epoch-hash");
+
+  private static final String EPOCH_HASH_FILE_SEPARATOR = "=";
+
+  private static final String EPOCH_HASH_FILE_READ_PATH = File.separator + "epoch-hash";
+
+  private static final byte[] LAST_EPOCH_VERIFIED_KEY =
+      "last_epoch_verified".getBytes(StandardCharsets.UTF_8);
+
+  public static final String ERROR_WHEN_OPENING_THE_EPOCHS_HASH_FILE =
+      "Error when opening the epochs hash file.";
 
   private Database substateAccumulatorHashDatabase;
   private Database epochHashDatabase;
+  private Database lastEpochHashVerifiedDatabase;
 
   private byte[] currentSubstateAccumulatorHash;
   private Optional<Long> lastEpochInDbOpt;
   private Optional<Long> lastEpochInFileOpt;
   private Optional<Long> lastStateVersionInDbOpt;
+  private Optional<Long> lastEpochHashVerified;
```

Comment:
Perhaps it worth to consider use of `OptionalLong`.

Reply:
Oh thanks for the suggestion, I didn't know it. I will use it in the future, but does it offer any big advantage to justify changing it now?

Reply:
It eliminates boxing/unboxing and not susceptible to NPE. If highest possible performance is not a requirement, then there is no significant difference between `OptionalLong` and `Optional<Long>`.